### PR TITLE
feat (api/hatchery/worker): worker model permission

### PR DESCRIPTION
**Pull request by fsamin**

### CODE REVIEW

File: engine/api/worker/worker.go [hunk 1]
```diff
@@ -226,55 +227,84 @@ type RegistrationForm struct {
...
-	if err != nil {
-		return nil, err
+	// Load token
+	t, errL := LoadToken(db, key)
+	if errL != nil {
+		return nil, errL
 	}
-	groupID = t.GroupID
-	e = t.Expiration
+	//groupID = t.GroupID
```
Comment (yesnault) [inline engine/api/worker/worker.go]: comment

File: sdk/worker.go [hunk 2]
```diff
@@ -39,21 +36,20 @@ var (
...
+	Capabilities []Requirement `json:"capabilities" db:"-"`
+	CreatedBy    User          `json:"created_by" db:"-"`
+	GroupID      int64         `json:"group_id" db:"group_id"`
 }
 
 // ModelStatus sums up the number of worker deployed and wanted for a given model
 type ModelStatus struct {
 	ModelID       int64         `json:"model_id" yaml:"-"`
 	ModelName     string        `json:"model_name" yaml:"name"`
+	ModelGroupID  int64         `json:"model_group_id" yaml:"name"`
```
Comment (yesnault) [inline sdk/worker.go]: yamp = model_group_id

File: vendor/vendor.json [hunk 3]
```diff
@@ -166,6 +166,12 @@
 			"revisionTime": "2016-10-24T02:17:13Z"
 		},
 		{
+			"checksumSHA1": "sd74nfsZXvigcGBf1gDXYzO2AKY=",
```
Comment (yesnault) [inline vendor/vendor.json]: To Delete gopkg.in/gorp.v1

File: engine/api/group/group.go [hunk 4]
```diff
@@ -186,6 +202,61 @@ func LoadGroupByUser(db database.Querier, userID int64) ([]sdk.Group, error) {
...
+	rows, err := db.Query(query, SharedInfraGroup)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var id int64
+		var name string
+		rows.Scan(&id, &name)
```
Comment (sguiheux) [inline engine/api/group/group.go]: missing err

File: engine/api/group/group.go [hunk 5]
```diff
@@ -186,6 +202,61 @@ func LoadGroupByUser(db database.Querier, userID int64) ([]sdk.Group, error) {
...
+	rows, err := db.Query(query, userID)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var id int64
+		var name string
+		rows.Scan(&id, &name)
```
Comment (sguiheux) [inline engine/api/group/group.go]: err ?

File: engine/api/scheduler/scheduler.go [hunk 6]
```diff
@@ -85,34 +83,34 @@ func PipelineScheduler(db *sql.DB, pb sdk.PipelineBuild) {
 	var runningStage = -1
 	var doneStage = 0
 	for stageIndex, s := range pb.Pipeline.Stages {
-		log.Info("PipelineScheduler> Pipeline %s #%d has stage %s (%d/%d)\n",
-			pb.Pipeline.Name,
-			pb.BuildNumber,
-			s.Name,
-			stageIndex+1,
-			len(pb.Pipeline.Stages))
+		/*log.Info("PipelineScheduler> Pipeline %s #%d has stage %s (%d/%d)\n",
```
Comment (sguiheux) [inline engine/api/scheduler/scheduler.go]: remove

File: engine/api/scheduler/scheduler.go [hunk 7]
```diff
@@ -85,34 +83,34 @@ func PipelineScheduler(db *sql.DB, pb sdk.PipelineBuild) {
...
-			log.Info("PipelineScheduler> Pipeline %s #%d (stage %d)[enabled=%v] has action %s [%s] (%d/%d)\n",
-				pb.Pipeline.Name,
-				pb.BuildNumber,
-				stageIndex,
-				s.Enabled,
-				a.Name,
-				status,
-				i+1,
-				len(s.Actions))
+			/*log.Info("PipelineScheduler> Pipeline %s #%d (stage %d)[enabled=%v] has action %s [%s] (%d/%d)\n",
```
Comment (sguiheux) [inline engine/api/scheduler/scheduler.go]: remove

File: engine/api/scheduler/scheduler.go [hunk 8]
```diff
@@ -159,16 +157,16 @@ func PipelineScheduler(db *sql.DB, pb sdk.PipelineBuild) {
 						continue
 					}
 				}
-
-				log.Debug("PipelineScheduler> %s.%s #%d has status %s\n", pb.Pipeline.Name, a.Name, pb.BuildNumber, status)
-
+				/*
```
Comment (sguiheux) [inline engine/api/scheduler/scheduler.go]: remove

File: engine/api/testwithdb/test.go [hunk 9]
```diff
@@ -13,12 +13,14 @@ import (
 	"testing"
 	"time"
 
-	"github.com/ovh/cds/engine/api/artifact"
+	"bytes"
```
Comment (sguiheux) [inline engine/api/testwithdb/test.go]: wrong place

File: engine/api/user_test.go [hunk 10]
```diff
@@ -1,14 +1,23 @@
...
 
 import (
+	"net/http/httptest"
 	"testing"
 
+	"github.com/gorilla/mux"
 	"github.com/proullon/ramsql/engine/log"
+	"github.com/stretchr/testify/assert"
 
+	"encoding/json"
```
Comment (sguiheux) [inline engine/api/user_test.go]: wrong place

File: engine/api/worker/model.go [hunk 11]
```diff
@@ -175,19 +163,10 @@ func LoadWorkerModelCapabilities(db database.Querier, workerID int64) ([]sdk.Req
 	var capas []sdk.Requirement
 	for rows.Next() {
 		var c sdk.Requirement
-		var typeS string
-		err = rows.Scan(&c.Name, &typeS, &c.Value)
+		err = rows.Scan(&c.Name, &c.Type, &c.Value)
```
Comment (sguiheux) [inline engine/api/worker/model.go]: shadow  => one line

File: engine/hatchery/swarm/swarm.go [hunk 12]
```diff
@@ -412,28 +412,52 @@ func (h *HatcherySwarm) killAwolWorker() {
...
 		}
 	}
+
+	//Delete the workers
+	for _, c := range oldContainers {
+		h.killAndRemove(c.ID)
+		log.Notice("HatcherySwarm.killAwolWorker> Delete worker %s\n", c.Names[0])
+	}
+
+	containers, err = h.dockerClient.ListContainers(docker.ListContainersOptions{
```
Comment (sguiheux) [inline engine/hatchery/swarm/swarm.go]: errLC ?

File: engine/sql/create_index.sql [hunk 13]
```diff
@@ -101,15 +101,13 @@ select create_index('user_key','IDX_USER_KEY_USER_KEY','user_key');
...
 select create_index('worker','IDX_WORKER_ID','id');
-select create_index('worker','IDX_WORKER_OWNER_ID','owner_id');
 
 -- WORKER_CAPABILITY
 select create_unique_index('worker_capability','IDX_WORKER_CAPABILITY_NAME','worker_model_id,name');
 select create_index('worker_capability','IDX_WORKER_CAPABILITY_MODEL_ID','worker_model_id');
 
 -- WORKER_MODEL
 select create_unique_index('worker_model','IDX_WORKER_MODEL_NAME','name');
-select create_index('worker_model','IDX_WORKER_MODEL_OWNER_ID','owner_id');
```
Comment (sguiheux) [inline engine/sql/create_index.sql]: So create an indexo n group_id ?

**Pull request merged**